File: week-5/course-schedule-ii/course-schedule-ii.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function(numCourses, prerequisites) {
    const graph = new Array(numCourses).fill(null).map(i=>[]);
    const visited = new Array(numCourses).fill(false);
    
    for (let [x, y] of prerequisites) {
        graph[x].push(y);
    }
    
    const res = [];
    
    for (let i=0; i<numCourses; i++) {
        if (!dfs(i, graph, visited, res)) return [];
    }
    
    return res;
};

const dfs = (i, graph, visited, res) => {
    if (visited[i] === -1) return false;
    if (visited[i]) return true;
    
    visited[i] = -1;
    for (let j of graph[i]) {
        if (!dfs(j, graph, visited, res)) return false;
    }
    
    res.push(i);
    visited[i] = true;
    return true;
}

File: week-5/course-schedule-ii/course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = [[] for i in range(numCourses)]
        visited = [False for i in range(numCourses)]
        
        for x, y in prerequisites:
            graph[x].append(y)
            
        res = []
        for i in range(numCourses):
            if not self.dfs(i, graph, visited, res):
                return []
            
        return res
    
    def dfs(self, i, graph, visited, res):
        if visited[i] == -1:
            return False
        
        if visited[i]: return True
        
        visited[i] = -1
        
        for j in graph[i]:
            if not self.dfs(j, graph, visited, res):
                return False
        
        visited[i] = True
        res.append(i)
        
        return True
